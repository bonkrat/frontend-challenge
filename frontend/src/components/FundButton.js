import React from "react";
import PropTypes from "prop-types";
import * as styles from "./FundButton.styles";

/**
 * Renders a button to fund a campaign/
 */
const FundButton = ({ loading, onClick }) => {
  return (
    <div
      className={styles.button(loading)}
      onClick={() => !loading && onClick()}
    >
      {loading ? "..." : "Fund"}
    </div>
  );
};

FundButton.defaultProps = {
  loading: false,
  onClick: () => {}
};

FundButton.propTypes = {
  /**
   * Whether the button should be inactive and loading.
   */
  loading: PropTypes.bool,
  /**
   * Handler for whent he button is clicked.
   */
  onClick: PropTypes.func
};

export default FundButton;
