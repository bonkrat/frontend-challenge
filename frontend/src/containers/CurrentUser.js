import { useState } from "react";
import { createContainer } from "unstated-next";

export const CurrentUser = () => {
  const [user, setUser] = useState({});
  return { user, setUser };
};

/**
 * Creates a global state container with the current logged in user.
 */
export default createContainer(CurrentUser);
