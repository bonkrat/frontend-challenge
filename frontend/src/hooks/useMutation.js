import { useMutation } from "@apollo/react-hooks";
import GlobalLoading from "../containers/GlobalLoading";
import { uniq, concat } from "lodash";
import { useEffect } from "react";
import { toast } from "react-toastify";

/**
 * A wrapper for useMutation that sets the global default loading state for any
 * component using this hook.
 */
export default (namespace, ...args) => {
  const { globalLoading, setGlobalLoading } = GlobalLoading.useContainer();
  const [mutation, { loading, error, data }] = useMutation(...args);

  if (error) {
    toast.error("Something went wrong! :(");
  }

  useEffect(() => {
    if (loading) {
      const newState = uniq(concat(globalLoading, [namespace]));
      setGlobalLoading(() => [...newState]);
    } else {
      setGlobalLoading(() =>
        globalLoading.filter(loadNamespace => loadNamespace !== namespace)
      );
    }
  }, [loading]);

  return [mutation, { loading, error, data }];
};
