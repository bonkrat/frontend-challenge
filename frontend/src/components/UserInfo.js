import React from "react";
import PropTypes from "prop-types";
import * as styles from "./UserInfo.styles";
import { USER_INFO_QUERY } from "./UserInfo.graphql";
import useQuery from "../hooks/useQuery";
import CurrentUser from "../containers/CurrentUser";
import ReactTooltip from "react-tooltip";

/**
 * Renders the logged in user's info
 */
export const UserInfo = ({ loading, name = "", avatar, balance }) => (
  <div className={styles.userInfo}>
    {balance && <div className={styles.balance}>${balance}</div>}
    <ReactTooltip effect="solid" place="bottom" />
    <div
      data-tip={name}
      className={styles.avatar(loading)}
      style={{
        backgroundImage: `url("http://localhost:4000/avatars/${avatar}")`,
        backgroundPosition: "center",
        backgroundSize: "cover"
      }}
    ></div>
  </div>
);

UserInfo.propTypes = {
  /**
   * Renders plcaeholder if loading
   */
  loading: PropTypes.bool,
  /**
   * The user's name
   */
  name: PropTypes.string,
  /**
   * URL to the users avatar
   */
  avatar: PropTypes.string,
  /**
   * The user's balance
   */
  balance: PropTypes.number
};

/**
 * Queries and renders the User's info.
 */
export const UserInfoQuery = () => {
  const { loading, error, data = {} } = useQuery("USER_INFO", USER_INFO_QUERY);
  const { setUser } = CurrentUser.useContainer();

  if (error) return <div></div>;
  setUser(data.loggedInUser);

  return <UserInfo loading={loading} {...data.loggedInUser} />;
};

export default UserInfoQuery;
