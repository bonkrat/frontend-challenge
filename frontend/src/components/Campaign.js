import React from "react";
import PropTypes from "prop-types";
import * as styles from "./Campaign.styles";
import useQuery from "../hooks/useQuery";
import { CAMPAIGN_QUERY } from "./Campaign.graphql";
import FundedBar from "./FundedBar";
import FundForm from "./FundForm";
import Backers from "./Backers";
import CampaignPlaceholder from "./CampaignPlaceholder";

/**
 * Renders a fullpage view of a campaign.
 */
export const Campaign = ({
  id,
  title,
  creator,
  image,
  goal,
  funds,
  description,
  backers
}) => (
  <React.Fragment>
    <div className={styles.campaignHeader}>
      <div className={styles.campaignTitle}>{title}</div>
      <div className={styles.campaignCreator}>Created by {creator.name}</div>
    </div>
    <div className={styles.campaignBody}>
      <img
        alt={title}
        className={styles.image}
        src={`http://localhost:4000/images/${image}`}
      />
      <div className={styles.campaignDetails}>
        <FundedBar className={styles.fundBar} funds={funds} goal={goal} />
        <div className={styles.goal}>
          <span className={styles.funds}>{funds < goal && `$${funds}`}</span>
          <span>${funds < goal ? goal : funds}</span>
        </div>
        <p className={styles.campaignDescription}>{description}</p>
        {backers.length > 0 && (
          <div className={styles.backedBy}>
            <div className={styles.backedByText}>Backed by:</div>
            <Backers backers={backers} size={12} />
          </div>
        )}
        <FundForm campaignId={id} />
      </div>
    </div>
  </React.Fragment>
);

Campaign.propTypes = {
  /**
   * The id of the campaign.
   */
  id: PropTypes.string,
  /**
   * The title of the campaign.
   */
  title: PropTypes.string,
  /**
   * The creator of the campaign
   */
  creator: PropTypes.object,
  /**
   * The URL of the image of the campaign.
   */
  image: PropTypes.string,
  /**
   * The goal to reach for the campaign.
   */
  goal: PropTypes.number,
  /**
   * The funds the campaign has raised.
   */
  funds: PropTypes.number,
  /**
   * The description of the campaign.
   */
  description: PropTypes.string,
  /**
   * The users that have backed the campaign.
   */
  backers: PropTypes.array
};

/**
 * Queries for and renders a campaign.
 */
export const CampaignQuery = ({ id }) => {
  const { loading, error, data: { campaign } = {} } = useQuery(
    "CAMPAIGN_VIEW",
    CAMPAIGN_QUERY,
    {
      variables: {
        id
      }
    }
  );

  if (error || loading) return <CampaignPlaceholder />;

  return <Campaign {...campaign} />;
};

CampaignQuery.propTypes = {
  /**
   * The id of the campaign to query and render.
   */
  id: PropTypes.string
};

export default CampaignQuery;
