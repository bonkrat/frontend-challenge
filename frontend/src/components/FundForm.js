import React, { useState } from "react";
import PropTypes from "prop-types";
import FundButton from "./FundButton";
import * as styles from "./FundForm.styles";
import { FUND_CAMPAIGN } from "./FundForm.graphql";
import useMutation from "../hooks/useMutation";
import CurrentUser from "../containers/CurrentUser";
import { CAMPAIGN_QUERY } from "./Campaign.graphql";
import { USER_INFO_QUERY } from "./UserInfo.graphql";
import { toast } from "react-toastify";

export const contributionOptions = [1, 5, 10, 25, 50, 100];

/**
 * Renders a form of contribution amounts and a button to contribute to a campaign.
 */
export const FundForm = ({
  fundCampaign,
  campaignId,
  userId,
  loading,
  selectedAmount,
  setAmount
}) => {
  return (
    <div className={styles.form}>
      <div className={styles.amounts}>
        {contributionOptions.map(amount => (
          <div
            id={`${amount}-button`}
            key={amount}
            className={styles.contributionAmountStyles(
              amount === selectedAmount
            )}
            onClick={() => {
              setAmount(amount);
            }}
          >
            {amount}
          </div>
        ))}
      </div>
      <FundButton
        onClick={() =>
          fundCampaign({
            variables: { campaignId, userId, amount: selectedAmount }
          })
        }
        loading={loading}
      />
    </div>
  );
};

FundForm.propTypes = {
  /**
   * Handler for funding a campaign.
   */
  fundCampaign: PropTypes.func,
  /**
   * The id of the campaign to fund.
   */
  campaignId: PropTypes.string,
  /**
   * The userId providing the funds.
   */
  userId: PropTypes.string,
  /**
   * True if loading.
   */
  loading: PropTypes.bool,
  /**
   * The current amount to fund that is selected.
   */
  selectedAmount: PropTypes.number,
  /**
   * Handler for setting the amount to fund.
   */
  setAmount: PropTypes.func
};

/**
 * Updates the cache after the mutation is completed.
 *
 * @param {string} campaignId  The ID of the campaign to fund.
 * @param {number} amount The amount of currency to fund the campaign with.
 */
export const updateCache = (campaignId, amount) => (cache, { data }) => {
  const { campaign } = cache.readQuery({
    query: CAMPAIGN_QUERY,
    variables: { id: campaignId }
  });

  const { loggedInUser } = cache.readQuery({
    query: USER_INFO_QUERY
  });

  cache.writeQuery({
    query: CAMPAIGN_QUERY,
    variables: { id: campaignId },
    data: {
      campaign: {
        ...campaign,
        ...data.fundCampaign
      }
    }
  });

  cache.writeQuery({
    query: USER_INFO_QUERY,
    data: {
      loggedInUser: {
        ...loggedInUser,
        balance: loggedInUser.balance - amount
      }
    }
  });
};

/**
 * Provides a mutation to fund a campaign.
 */
export const FundFormMutation = ({ campaignId }) => {
  const [selectedAmount, setAmount] = useState(1);
  const [fundCampaign, { loading }] = useMutation("FUND_FORM", FUND_CAMPAIGN, {
    update: updateCache(campaignId, selectedAmount),
    onCompleted: () => {
      toast.success("You successfully funded this campaign!", {
        position: toast.POSITION.BOTTOM_RIGHT
      });
    }
  });
  const { user } = CurrentUser.useContainer();

  return (
    <FundForm
      campaignId={campaignId}
      userId={user && user.id}
      fundCampaign={fundCampaign}
      loading={loading}
      selectedAmount={selectedAmount}
      setAmount={setAmount}
    />
  );
};

FundFormMutation.propTypes = {
  /**
   * The id of the campaign to fund.
   */
  campaignId: PropTypes.string
};

export default FundFormMutation;
