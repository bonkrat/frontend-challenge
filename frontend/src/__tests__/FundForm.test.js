import React, { useState } from "react";
import { shallow } from "enzyme";
import {
  FundFormMutation,
  FundForm,
  updateCache,
  contributionOptions
} from "../components/FundForm";
import toJson from "enzyme-to-json";
import FundButton from "../components/FundButton";
import useQuery from "../hooks/useQuery";
import useMutation from "../hooks/useMutation";
import { FUND_CAMPAIGN } from "../components/FundForm.graphql";
import { toast } from "react-toastify";
import CurrentUser from "../containers/CurrentUser";
import { CAMPAIGN_QUERY } from "../components/Campaign.graphql";
import { USER_INFO_QUERY } from "../components/UserInfo.graphql";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(() => [1, jest.fn()])
}));

jest.mock("../hooks/useMutation", () =>
  jest.fn(() => [jest.fn(), { loading: false }])
);

jest.mock("../containers/CurrentUser", () => ({
  useContainer: jest.fn(() => ({
    user: {
      id: "userfoo",
      balance: 500
    }
  }))
}));

jest.mock("react-toastify", () => ({
  toast: {
    success: jest.fn(),
    POSITION: {
      BOTTOM_RIGHT: "br"
    }
  }
}));

describe("FundForm", () => {
  const mockFundFn = jest.fn();
  const mockSetFn = jest.fn();
  let wrapper;

  const props = {
    fundCampaign: mockFundFn,
    campaignId: "foo",
    userId: "foo",
    loading: false,
    selectedAmount: 5,
    setAmount: mockSetFn
  };

  beforeEach(() => {
    wrapper = shallow(<FundForm {...props} />);
  });

  it("renders correctly", () => {
    shallow(<FundForm {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders loading styles correctly", () => {
    wrapper.setProps({
      ...props,
      loading: true
    });
    wrapper.update();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("sets the amount when an amount is clicked", () => {
    contributionOptions.forEach(amount => {
      const handler = wrapper
        .findWhere(n => n.prop("id") === `${amount}-button`)
        .prop("onClick");

      handler();

      expect(props.setAmount).toHaveBeenCalledWith(amount);
    });
  });

  it("funds the campaign when the button is clicked", () => {
    const handler = wrapper.find(FundButton).prop("onClick");
    handler();
    expect(props.fundCampaign).toHaveBeenCalledWith({
      variables: {
        campaignId: props.campaignId,
        userId: props.userId,
        amount: 5
      }
    });
  });
});

describe("FundFormMutation", () => {
  let wrapper;

  const props = {
    campaignId: "foo"
  };

  beforeEach(() => {
    wrapper = shallow(<FundFormMutation {...props} />);
  });

  afterEach(() => {
    useState.mockClear();
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("passes the selected amount from state", () => {
    expect(useState).toHaveBeenCalledWith(1);

    expect(wrapper.find(FundForm).prop("selectedAmount")).toBe(1);

    wrapper.find(FundForm).prop("setAmount")();

    expect(useState.mock.results[0].value[1]).toHaveBeenCalled();
  });

  it("uses a the fund form mutation", () => {
    expect(useMutation).toHaveBeenCalled();
    expect(useMutation.mock.calls[0][0]).toBe("FUND_FORM");
    expect(useMutation.mock.calls[0][1]).toBe(FUND_CAMPAIGN);
    expect(useMutation.mock.calls[0][2].update.toString()).toEqual(
      updateCache("foo", 1).toString()
    );
  });

  it("calls a toast on completion", () => {
    useMutation.mock.calls[0][2].onCompleted();
    expect(toast.success).toHaveBeenCalledWith(
      "You successfully funded this campaign!",
      {
        position: "br"
      }
    );
  });

  it("provides the user from global state", () => {
    expect(CurrentUser.useContainer).toHaveBeenCalled();
    expect(wrapper.find(FundForm).prop("userId")).toBe("userfoo");
  });
});

describe("updateCache", () => {
  it("writes the new mutation data to the cache", () => {
    const updateFn = updateCache("foo", 123);

    const cache = {
      readQuery: jest.fn(() => ({
        loggedInUser: {
          id: "foouser",
          balance: 124
        }
      })),
      writeQuery: jest.fn()
    };
    const data = {
      data: {
        campaign: {
          id: "foocampaign"
        },
        fundCampaign: {
          title: "footitle"
        }
      }
    };
    updateFn(cache, data);

    expect(cache.readQuery).toHaveBeenCalledTimes(2);
    expect(cache.readQuery).toHaveBeenCalledWith({
      query: CAMPAIGN_QUERY,
      variables: {
        id: "foo"
      }
    });
    expect(cache.readQuery).toHaveBeenCalledWith({
      query: USER_INFO_QUERY
    });
    expect(cache.writeQuery).toHaveBeenCalledTimes(2);
    expect(cache.writeQuery).toHaveBeenCalledWith({
      query: CAMPAIGN_QUERY,
      variables: { id: "foo" },
      data: {
        campaign: {
          title: "footitle"
        }
      }
    });
    expect(cache.writeQuery).toHaveBeenCalledWith({
      query: USER_INFO_QUERY,
      data: {
        loggedInUser: {
          id: "foouser",
          balance: 1
        }
      }
    });
  });
});
