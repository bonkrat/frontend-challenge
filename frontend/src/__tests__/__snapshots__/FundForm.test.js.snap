// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FundForm renders correctly 1`] = `
<div
  className="flex flex-auto lg:justify-start lg:px-0 md:pr-48 sm:justify-between items-end"
>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="1-button"
    key="1"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer bg-purple-800 text-white"
    id="5-button"
    key="5"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="10-button"
    key="10"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="25-button"
    key="25"
    onClick={[Function]}
  >
    25
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="50-button"
    key="50"
    onClick={[Function]}
  >
    50
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="100-button"
    key="100"
    onClick={[Function]}
  >
    100
  </div>
  <FundButton
    loading={false}
    onClick={[Function]}
  />
</div>
`;

exports[`FundForm renders loading styles correctly 1`] = `
<div
  className="flex flex-auto lg:justify-start lg:px-0 md:pr-48 sm:justify-between items-end"
>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="1-button"
    key="1"
    onClick={[Function]}
  >
    1
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer bg-purple-800 text-white"
    id="5-button"
    key="5"
    onClick={[Function]}
  >
    5
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="10-button"
    key="10"
    onClick={[Function]}
  >
    10
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="25-button"
    key="25"
    onClick={[Function]}
  >
    25
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="50-button"
    key="50"
    onClick={[Function]}
  >
    50
  </div>
  <div
    className="bg-gray-200 rounded mx-1 md:px-4 px-4 py-2 cursor-pointer"
    id="100-button"
    key="100"
    onClick={[Function]}
  >
    100
  </div>
  <FundButton
    loading={true}
    onClick={[Function]}
  />
</div>
`;

exports[`FundFormMutation renders correctly 1`] = `
<FundForm
  campaignId="foo"
  fundCampaign={[MockFunction]}
  loading={false}
  selectedAmount={1}
  setAmount={[MockFunction]}
  userId="userfoo"
/>
`;
